### ISSUE TITLE
Refine upgrade pathway to z2jh 4 (JupyterHub 5, KubeSpawner 7, OAuthenticator 17)

### ISSUE THREAD

**Issue by consideRatio** [2i2c-org/infrastructure#5079]
Z2JH 4.0.0 is now out ([changelog](https://z2jh.jupyter.org/en/latest/changelog.html)). This issue tracks refining work to make it get done.

I've already trialed upgrading z2jh from 3 to 4 (beta-4) for a 2i2c deployment (https://github.com/2i2c-org/infrastructure/issues/4916, https://github.com/2i2c-org/infrastructure/pull/4979), and I figure its reasonable for 2i2c to upgrade all hubs to z2jh 4.0.0 at this point in time.

---

## Tasks

- [x] verify chart changelog breaking changes
- [x] verify sub-projects new versions breaking changes
- [x] verify compatibility with user images not using jupyterhub 5

## Definition of done

- Other issues have been created and refined to drive us towards upgrading to z2jh 4 more concretely.

**Comment by yuvipanda** [comment]
We would need https://github.com/yuvipanda/jupyterhub-fancy-profiles/pull/72 merged and deployed as well before we can upgrade

**Comment by GeorgianaElena** [comment]
## TL;DR

I just went through all of the breaking changes mentioned by the chat's changelog and also checked the changelogs of the related sub-projects and basehub dependencies.

I am pasting below my findings. The checked boxes aren't impacting us, whereas the others have an action attached to them that we need to take on the 2i2c side.

The plan is open a planning issue and what's below will be turned into an ordered list of sub-issues

### Breaking changes that the chart brings:

- [x] The chart now require Kubernetes 1.28+, up from 1.23+
- [x] Python is upgraded in chart images from 3.11 to 3.12
- [ ] KubeSpawner is upgraded from 6.2.0 to 7.0.0

    - **Action**: per [the upgrade docs](https://jupyterhub-kubespawner.readthedocs.io/en/stable/templates.html#templates-upgrading-from-less-than-7 
        ), update volume mounts to 
    ```
        c.KubeSpawner.volume_mounts = {
            "name": "home",
            "mountPath": "/home/jovyan",
            "subPath": "{escaped_username}",  # matches "{username}" in kubespawner 6
        }
    ```

- [ ] JupyterHub 4.1.6 has been upgraded to 5.2.1

    - [ ] switch from jupyter-telemetry to jupyter-events:
        - **Action**:
            - change https://github.com/2i2c-org/infrastructure/blob/5bb90103eb68479a14957547a3c3459a6b835a62/helm-charts/basehub/values.yaml#L71
            - to EventLogger per https://github.com/jupyterhub/jupyterhub/pull/4807
    - [ ] update bootstrap to v5
        - **Action** https://github.com/2i2c-org/jupyterhub-fancy-profiles/issues/45
    - [x] explicitly require groups in auth model when Authenticator.manage_groups is enabled
    - [x] add JupyterHub.subdomain_hook: Only used when `JupyterHub.subdomain_host` is set and we're not setting it

- [ ] OAuthenticator 16.3.1 has been upgraded to 17.1.0

     - [ ] [Generic, OpenShift] OAuthenticator.allowed_groups and OAuthenticator.admin_groups now require OAuthenticator.manage_groups to be set to True, and will otherwise raise an error
        - **Action**: set manage_groups to True as we're configuring admin_groups pretty much for all hubs 
- [x] LDAPAuthenticator 1.3.2 has been upgraded to 2.0.2

       
Infrastructure dependencies related **actions**:
- [ ] Rm https://github.com/2i2c-org/infrastructure/blob/main/helm-charts/images/hub/pkce-requirements.txt, revert https://github.com/2i2c-org/infrastructure/pull/4968/files
- [ ] Check if jupyterhub-configurator pinned branch still works with latest z2jh chart

**Comment by GeorgianaElena** [comment]
The plan has been laid out in https://github.com/2i2c-org/infrastructure/issues/5209 and its sub-issues
